io/src/mapper.cpp

```diff
@@ -251,12 +289,34 @@ measurement_particle_map generate_measurement_particle_map(
     }
 
     // generate cell particle map
-    auto c_p_map = generate_cell_particle_map(event, cells_dir, hits_dir,
-                                              particle_dir, link_map);
-
-    for (auto const& [meas, cells] : m_c_map) {
-        for (const auto& c : cells) {
-            result[meas][c_p_map[c]]++;
+    auto p_c_map = generate_particle_cell_map(event, cells_dir, hits_dir,
+                                              particle_dir, resource, link_map);
+
+    // Loop over the map associating cells with measurements.
+    for (auto const& [meas, mCells] : m_c_map) {
+        // Loop over the map associating particles with cells.
+        for (auto const& [ptc, pCells] : p_c_map) {
+            // Check how many cells are shared between the two.
+            for (std::size_t mCellIndex = 0; mCellIndex < mCells.size();
+                 ++mCellIndex) {
+                for (std::size_t pCellIndex = 0; pCellIndex < pCells.size();
+                     ++pCellIndex) {
+                    if ((mCells.channel0().at(mCellIndex) ==
+                         pCells.channel0().at(pCellIndex)) &&
+                        (mCells.channel1().at(mCellIndex) ==
+                         pCells.channel1().at(pCellIndex)) &&
+                        (std::abs(mCells.activation().at(mCellIndex) -
+                                  pCells.activation().at(pCellIndex)) <
+                         1e-6f) &&
+                        (std::abs(mCells.time().at(mCellIndex) -
+                                  pCells.time().at(pCellIndex)) < 1e-6f) &&
+                        (mCells.module_index().at(mCellIndex) ==
+                         pCells.module_index().at(pCellIndex))) {
+
+                        result[meas][ptc]++;
+                    }
+                }
+            }
```

Comment:
Question: what is happening here? :worried:

Reply:
:laughing: I also hate this...

What happens here is that we have two maps. One associating cells with all the "truth measurements" in the event. The other associating cells with all the "truth particles" in the event. And here we try to figure out which "truth measurements" belong to which "truth particles".

Since in our data cells can appear multiple times, when particles hit the same part of a surface, I had to require identity between every variable of the cells here. At first I was only checking for their integer identifiers. And that made the unit tests fail. That's why this ugly comparison was made.

Without a question this could also be expressed a lot more nicely with proxy types...